internal/pkg/prune/prune.go

```diff
@@ -0,0 +1,248 @@
+			fmt.Fprintf(os.Stderr, "retrieving current configuration of %s from server for %v", u.GetName(), err)
+			return Result{o.Resources}, err
+		}
+		fmt.Fprintf(os.Stderr, "couldn't find a configmap for pruning %s", u.GetName())
+		return Result{}, nil
+	}
+
+	hash := u.GetAnnotations()["current"]
+	newObj, err := o.runPrune(hash, obj)
+	if err != nil {
+		return Result{o.Resources}, err
+	}
+
+	// update the configmap object
+	err = o.Client.Update(context.Background(), newObj)
+	if err != nil {
+		return Result{}, err
+	}
+	fmt.Fprintf(o.Out, "%s/%s updated\n", u.GetKind(), u.GetName())
+	return Result{Resources: o.Resources}, nil
+}
+
+func (o *Prune) runPrune(hash string, cm *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+
+	toKeep := make(map[string]string)
+	old := make(map[string]string)
+
+	refOther := make(map[string][]string)
+	refByOther := make(map[string][]string)
+
+	toDelete := []string{}
+
+	all := cm.Object["data"].(map[string]interface{})
+
+	//first round, keep all the items that have the same hash
+	for key, value := range all {
+		h := value.(string)
+		if h == hash {
+			toKeep[key] = h
+			delete(all, key)
+		} else if !strings.Contains(key, "---") {
+			old[key] = h
+		}
+
+		if strings.Contains(key, "---") {
+			objs := strings.Split(key, "---")
+			if len(objs) != 2 {
+				return nil, fmt.Errorf("not able to get 2 objects from %s", key)
+			}
+			if _, ok := refByOther[objs[0]]; ok {
+				refByOther[objs[0]] = append(refByOther[objs[0]], objs[1])
+			} else {
+				refByOther[objs[0]] = []string{objs[1]}
+			}
+
+			if _, ok := refOther[objs[1]]; ok {
+				refOther[objs[1]] = append(refOther[objs[1]], objs[0])
+			} else {
+				refOther[objs[1]] = []string{objs[0]}
+			}
+		}
+	}
+
+	for key := range old {
+		if _, ok := refByOther[key]; !ok {
+			toDelete = append(toDelete, key)
+			if _, ok2 := refOther[key]; ok2 {
+				delete(refOther, key)
+			}
+		}
+	}
+
+	for key, value := range old {
+		if refs, ok := refByOther[key]; ok {
+			found := false
+			for _, ref := range refs {
+				if isInMap(ref, toKeep) && o.shouldKeep(key, ref) {
+					toKeep[key] = value
+					toKeep[key+"---"+ref] = value
+					found = true
+				}
+			}
+			if !found {
+				toDelete = append(toDelete, key)
+			}
+		}
+	}
+
+	err := o.deleteObjects(toDelete)
+	if err != nil {
+		return nil, err
+	}
+
+	cm.Object["data"] = toKeep
+	return cm, nil
+}
+
+func isInMap(key string, m map[string]string) bool {
+	if _, ok := m[key]; ok {
+		return true
+	}
+	return false
+}
+
+func (o *Prune) shouldKeep(obj1, obj2 string) bool {
+	values := strings.Split(obj1, "_")
+	switch values[1] {
+	case "PersistentVolume":
+		return true
+	case "ConfigMap", "Secret":
+		switch v := strings.Split(obj2, "_"); v[1] {
+		case "Deployment", "StetefulSet":
+			return o.isUsedInDeployment(obj1, obj2)
+		default:
+			return false
+		}
+	default:
+		return false
+	}
+}
+
+func (o *Prune) deleteObjects(items []string) error {
+	for _, item := range items {
```

Comment:
as discussed offline - lets make this more structured with annotations instead of data.